tableauserverclient/server/endpoint/datasources_endpoint.py

```diff
@@ -173,22 +173,37 @@ def delete_extract(self, datasource_item):
     @api(version="2.0")
     @parameter_added_in(connections="2.8")
     @parameter_added_in(as_job='3.0')
-    def publish(self, datasource_item, file_path, mode, connection_credentials=None, connections=None, as_job=False):
-        if not os.path.isfile(file_path):
-            error = "File path does not lead to an existing file."
-            raise IOError(error)
-        if not mode or not hasattr(self.parent_srv.PublishMode, mode):
-            error = 'Invalid mode defined.'
-            raise ValueError(error)
+    def publish(self, datasource_item, file, mode, connection_credentials=None, connections=None, as_job=False):
+
+        try:
+
+            if not os.path.isfile(file):
+                error = "File path does not lead to an existing file."
+                raise IOError(error)
+
+            filename = os.path.basename(file)
+            file_extension = os.path.splitext(filename)[1][1:]
+            file_size = os.path.getsize(file)
 
-        filename = os.path.basename(file_path)
-        file_extension = os.path.splitext(filename)[1][1:]
+            # If name is not defined, grab the name from the file to publish
+            if not datasource_item.name:
+                datasource_item.name = os.path.splitext(filename)[0]
+            if file_extension not in ALLOWED_FILE_EXTENSIONS:
+                error = "Only {} files can be published as datasources.".format(', '.join(ALLOWED_FILE_EXTENSIONS))
+                raise ValueError(error)
 
-        # If name is not defined, grab the name from the file to publish
-        if not datasource_item.name:
-            datasource_item.name = os.path.splitext(filename)[0]
-        if file_extension not in ALLOWED_FILE_EXTENSIONS:
-            error = "Only {} files can be published as datasources.".format(', '.join(ALLOWED_FILE_EXTENSIONS))
+        except TypeError:
+
+            if not datasource_item.name:
+                error = "Datasource item must have a name when passing a file object"
+                raise ValueError(error)
+
+            file_extension = 'tdsx' if file_is_compressed(file) else 'tds'
```

Comment:
I think we might have a problem here -- you can directly publish a `tde` or `hyper` file, both of which are actually compressed file formats themselves (I'm not sure if they're both zip or some other format)

Reply:
This might not be an issue practically -- if they're loading a tde or hyper file in memory it might just fail on publish.

I wonder if there is a way to warn the caller early though.

Reply:
I modified the file_peeking to return file_type instead of boolean value and extended it to identify .xml, .tde, and .hyper. I couldn't find official sources for .tde or .hyper file signatures but was able to deduce them by creating few different files and peeking their content. 

Now if users provides '.tde' or '.hyper' file they are informed that these are not-supported file types. Additionally, they will be shown an error if they're providing file types we can't identify. Let me know what you think